### DISCUSSION TITLE
A problem about the setting of neutron boundary condition in Moltres simulation

### DISCUSSION THREAD

**Discussion by js-jixu** [arfc/moltres#196] [Q&A Reactor multiphysics] [answered]
Hello, everyone. I'm a new Moltres user and I've encountered with  a problem about the setting of neutron boundary condition in Moltres simulation. Any advice or suggestions are  greatly appreciated.

These days I try to replace only one part of the [tutorial input file](http://arfc.github.io/software/moltres/wiki/input_example/) with other things. After some attempts, I have found that changes of flow velocity of precursors and fuel salt, thermodynamic properties and boundary conditions of temperature will not affect the correctness of results, while the change of boundary condition of neutron will lead to incorrectness of results.

In the tutotial input file called "auto_diff_rho", located in " ~/projects/moltres/tests/twod_axi_coupled/ ", I just change neutron boundary conditions of the BCs block. I replace the vacuum bc with neumann bc while the value is 0. The complete content of BCs block of input file is as follows:

[BCs]
  [./vacuum_group1]
    type = NeumannBC
    boundary = 'fuel_bottoms fuel_tops moder_bottoms moder_tops outer_wall'
    variable = group1
    value = 0
  [../]
  [./vacuum_group2]
    type = NeumannBC
    boundary = 'fuel_bottoms fuel_tops moder_bottoms moder_tops outer_wall'
    variable = group2
    value = 0
  [../]
  [./temp_advection_outlet]
    boundary = 'fuel_tops'
    type = TemperatureOutflowBC
    variable = temp
    velocity = '0 ${flow_velocity} 0'
  [../]
  [./temp_diri_cg]
    boundary = 'moder_bottoms fuel_bottoms outer_wall'
    type = FunctionDirichletBC
    function = 'temp_bc_func'
    variable = temp
  [../]
[]

![all_reflection_MSR_group1](https://user-images.githubusercontent.com/99813361/163573511-48f8b815-e683-4e52-b069-7501ce719bd9.png)
![all_reflection_MSR_group2](https://user-images.githubusercontent.com/99813361/163573517-b3295007-b641-4d7e-9543-411c3a53c4a6.png)
![all_reflection_MSR_pre1](https://user-images.githubusercontent.com/99813361/163573518-7ceb1b18-33be-4167-ab03-d86a0a97ed87.png)
![all_reflection_MSR_temp](https://user-images.githubusercontent.com/99813361/163573521-4580a275-98e9-43d2-b757-2a1c9921e247.png)

I think vacuum boundary condition is similar with reflection boundary condition in a sense, but why are their results so different?

**Comment by smpark7** [comment, marked as answer]
Hi Xu. Thanks for making the first post here!

Vacuum BC and reflective BC are completely different! Vacuum BCs assume that there is no incoming flux. We typically use vacuum BCs on the outer boundaries of our reactor model. Reflective BCs assume there is no net flux on the boundary. We typically use reflective BCs to model infinite lattices or to exploit symmetries in our reactor model to reduce the problem size. 

In this case, when you changed the BCs to reflective, you eliminated neutron loss through leakage. The only stable solution for this problem is zero neutron flux everywhere or some pattern of unphysical positive and negative fluxes. Your image shows the patterns I mentioned. 

Regarding your LFR model we discussed over email, there is nothing you can do or test until you fix your mesh file. A broken mesh will give you unphysical results no matter what you change in your input file. Good luck!

**Reply by js-jixu** [reply]
Hi Sun Myung. Thanks a lot for answering my question. I still have three questions for the simulation.

For a reactor simulation, I think a total power is usually preset before the simulation, but I don't find any information about the preset total power in [the tutorial](http://arfc.github.io/software/moltres/wiki/input_example/). Can you tell me the setting of the power in the MSRE simulation?

I've remembered that you said that if the input file don't set a boundary condition for a boundary, MOOSE will set a NeumannBC for the boundary. Whether the value of the NeumannBC is 0?

I can understand that Reflective BC will give a steady plane neutron flux distribution, but I can't understand why the value of the plane neutron flux is zero. Can you tell me that neutron flux distribution is a relative distribution or an absolute distribution, I mean that whether the neutron flux is normalized?

I want to use the tutorial geometry and mesh to be  familiar with Moltres and MOOSE now. And I want to use gmsh or mesh-generator provided by MOOSE to generate a new LFR single rod model in next step. I would also like to ask you how did you learn gmsh in the first place. For someone who has never modeled and drawn meshes, do you recommend Gmsh or mesh-generator provided by MOOSE?

Thanks a lot again!

**Reply by js-jixu** [reply]
Hi Sun Myung. I find some new problems when I check the postprocessors generated by the tutorial input file with Neutron Reflective BC. The problem is the appearace of negative neutron flux and the absolute value of the negative neutron flux is big. The postprocessors generated in time step 11 is here.

![negative postprocessors](https://user-images.githubusercontent.com/99813361/164481631-8d1b145b-f4d1-4725-ad95-8e3acef299ac.png)

In the tutorial input file, I replaced the thermal properties of molten salt with that of the MOX fuel, replaced the thermal properties of the graphite with that of the lead-bismuth eutectic, and set the MOX fuel to no flow and the lead-bismuth eutectic coolant to flow. And I replace VacuumBC with ReflectiveBC in the original input file, while the geometry and mesh don't change. Correspondingly, the velocity of precursors equals to 0 beacuse there is no flow in fuel region. Here are results. Usually, when ReflectiveBC is set, where the neutron flux is relatively high, the temperature is usually lower, and where the neutron flux is relatively low, the temperature is usually higher. However, the trend of the following three pictures is not as obvious as the above three pictures. Do you think the results are right?

![lfr_fast_group](https://user-images.githubusercontent.com/99813361/164486506-a9a8ce67-a6dc-432b-867c-a191d802a3ce.png)
![lfr_thermal_group](https://user-images.githubusercontent.com/99813361/164486553-96a7e07d-23e9-4abf-b0a8-1b9ccfac64e2.png)
![lfr_temp](https://user-images.githubusercontent.com/99813361/164486591-ddf23f07-6755-40f1-ab83-66352ecb4411.png)
[lfr_reflective.txt](https://github.com/arfc/moltres/files/8532284/lfr_reflective.txt)

The LFR input file with ReflectiveBC is attached.

With best wishes!

**Comment by smpark7** [comment]
Hi Xu. There are many factors to consider in your simulations.
- The tutorial input file runs a transient (time-dependent) simulation in which we allow the reactor power to vary according to the reactor state (temperature distribution and cross sections). For a fixed power output simulation, we usually run an eigenvalue/criticality solve simulation which requires a different problem setup. I'm currently setting up another tutorial for that type of simulation. In the meantime, you may refer to `moltres/problems/2021-cnrs-benchmark/phase-0/nts.i` for an example of how to set this up. This input file simulates a 2D 200cm x 200cm molten salt region with the total power output set to 10MW.
- The default boundary condition is homogeneous Neumann BC, which is Neumann BC with the constant value set to zero.
- Your simulations with reflective BCs all failed to run correctly. Setting reflective BCs causes the initial reactor state to be very supercritical (k>1). In such a system, we expect to see a large prompt jump in neutrons. The prompt response typically occurs on the order of 10^-6 (0.000001) seconds. The initial timestep size of 10^-3 is too large to accurately simulate this system. Therefore, Moltres fails to converge properly, and it provides unphysical, negative flux results. This is also the reason for your negative fluxes when you changed the material composition.
- I would recommend Gmsh because it has more functionalities than MOOSE mesh generator. The MOOSE mesh generator is good for simple geometries (e.g. simple shapes, regular pin assemblies). Gmsh also has a more comprehensive tutorial guide than the MOOSE mesh generator.

I strongly recommend focusing on learning how to mesh simple geometries before experimenting with Moltres so that you're able to replicate the geometries of the systems you intend to simulate. Replacing material properties of a reactor design with those from another reactor design will not give you ideal results because of how sensitive the multiplication factor is to both geometry and cross sections.

**Reply by js-jixu** [reply]
Hi Sun Myung. Thanks a lot for your reply and your reply helps me a lot. Now I'm learning the usage of gmsh. I've found a PDF translate website, which has translated English gmsh manual into Chinese one.

And I run a simulation of the tutorial input file with Reflective BC and a small time step. What I want to confirm is whether Moltres can simulate Reflective BC. The initial time step I used is 10-8s and the maxmium of time step is 10-3s.  Do you think the time step is reasonable?
[msr_reflective.txt](https://github.com/arfc/moltres/files/8548922/msr_reflective.txt)

With best wishes.

**Reply by smpark7** [reply]
I would say that the initial time step is too small. The simulation will take very long to show any meaningful change in your variables. I recommend starting with dt=10^-6s. As for the maximum timestep, I think it's okay to not specify any limit for now. The IterativeAdaptiveDT automatically adjusts the timestep according to the number of iterations needed per timestep. If you find that your timestep keeps getting cut back because the timestep increase is too aggressive, you can set a maximum to prevent it from reaching too large timesteps or adjust the `growth_factor` or `optimal_iterations`.

**Reply by js-jixu** [reply]
Thanks you a lot, Sun Myung.

I‘ve tried bigger time step, such as dt=10^-6s and dtmax=10-3s, but I finally get nagative neutron flux. Now I don't bother with this any more since my boss told me that Moltres had been verified, so it should be fine.

I have learned the usage of gmsh these days and constructed a LFR single rod input file with mesh file generated by gmsh again. But a new problem about mesh file appears. I'll put it in a [new discussion](https://github.com/arfc/moltres/discussions/198).

With best wishes, Xu.
